File: 77-Combinations-组合/77-Combinations-组合.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
 let res = []
 let path = []
 var combine = function(n, k) {
     res = []                    // 如果要定义全局变量，则每次测试时要重置，因为leetcode是从一个文件测试所有的案例。
     backtrack(n, k, 1)
     return res
 };
 
 var backtrack = function(n, k, index){
     if(path.length === k){
         res.push(Array.from(path))
         return
     }
 
     for(let i = index; i <= n - (k - path.length) + 1; i++){
         path.push(i)
         backtrack(n, k, i + 1)
         path.pop()
     }
 }

File: 77-Combinations-组合/77-Combinations-组合.java
class Solution {
    
    private List<List<Integer>> res;
    private LinkedList<Integer> path = new LinkedList<>();
    
    public List<List<Integer>> combine(int n, int k) {
    	res = new ArrayList<>();
    	backtrace(n, k, 1);
    	return res;
    }
    
    public void backtrace(int n, int k, int index) {
        // 每次递归表示第K个数
    	if(path.size() == k) {
    		res.add(new ArrayList<Integer>() {{addAll(path);}});
            return;
    	}
    	
        // 每次循环表示第k个数可能的数的遍历，注意不能重复，所以起始是从上一次的值开始
    	for(int i = index; i <= n - (k - path.size()) + 1; i++) {
    		path.add(i);
    		backtrace(n, k, i + 1);
    		path.removeLast();
    	}
    }
}